### ISSUE TITLE
TrainableTokensModel doesn't handle embedding scaling

### ISSUE THREAD

**Issue by kimihailv** [huggingface/peft#2809]
### System Info

peft: 0.17.1
transformers: 4.56.2

### Who can help?

@BenjaminBossan

hello, I want to fine-tune some embeddings of Gemma 3. Gemma 3 uses `Gemma3TextScaledWordEmbedding` inherited from `torch.nn.Embedding`. The forward method of this class [scales](https://github.com/huggingface/transformers/blob/9d8f693c7ece66c69459bc5db794234fb26df5d3/src/transformers/models/gemma3/modeling_gemma3.py#L107) the output of embedding layer by some constant (it is stored in model's buffer). The problem is that `TrainableTokensLayer` [relies](https://github.com/huggingface/peft/blob/815956b9b85f893e5acd2dd1d8391954b0f93e07/src/peft/tuners/trainable_tokens/layer.py#L225) on `isinstance` method which can't distinguish between class and its children and because of it skips the custom forward logics of `Gemma3TextScaledWordEmbedding`. This leads to high gradients norm and loss.

### Reproduction

```
from transformers import AutoModelForCausalLM, AutoTokenizer
from peft import TrainableTokensModel, TrainableTokensConfig
import torch

device = "cuda:0"
model1 = AutoModelForCausalLM.from_pretrained("google/gemma-3-270m", device_map=device)
model2 = AutoModelForCausalLM.from_pretrained("google/gemma-3-270m", device_map=device)
tokenizer = AutoTokenizer.from_pretrained("google/gemma-3-270m")

config = TrainableTokensConfig(
        token_indices=[1, 2],
        init_weights=True,
    )

model2 = TrainableTokensModel(model2, config, "trainable_tokens")

x = torch.tensor([[1, 2]], device=device, dtype=torch.long)

y1 = model1.model.embed_tokens(x)
y2 = model2.model.model.embed_tokens(x)

print("src embeddings:", y1)
print("embeddings with adapter:", y2)

print("scaled embeddings:", y2 * model1.model.embed_tokens.embed_scale)
```

### Expected behavior

Adapter should distinguish between vanilla torch.nn.Embedding layer and custom variants.

**Comment by BenjaminBossan** [comment]
Thanks for the bug report. It is indeed true that the embedding scaling is not taken into account by PEFT. This is not only true for trainable tokens, but also for other PEFT methods like LoRA.

> Adapter should distinguish between vanilla torch.nn.Embedding layer and custom variants.

I don't like this solution too much. There could be subclasses of `nn.Embedding` that are perfectly fine to be adopted. If we disallowed that, a lot of valid code could stop working.

Instead, I'm considering to add explicit support for this to PEFT. When I look at the [`Gemma3TextScaledWordEmbedding` implementation](https://github.com/huggingface/transformers/blob/bcdd5532bfd48660c029a0ddb8f9c76cb02039b2/src/transformers/models/gemma3/modeling_gemma3.py#L97), it should be quite simple. However, I also wouldn't want to add an extra implementation just for Gemma3 embeddings. So an idea I had was to check for the `self.embed_scale` attribute and if it exists (and has the right shape), we apply this scale in the PEFT classes that work on embeddings. Although not perfect, I think it would still work in most cases and be more generally applicable. WDYT?

**Comment by kimihailv** [comment]
I think it is a reasonable solution if you don't want to be too restrictive. But of course there are no guarantees that a scale constant is always named "embed_scale", and also some can scale by multiplication, while other can use division...

**Comment by BenjaminBossan** [comment]
> But of course there are no guarantees that a scale constant is always named "embed_scale", and also some can scale by multiplication, while other can use division...

Of course, the proposed solution would rely purely on convention and could be missing similar instances that use a different variable name. However, after grepping through the transformers code base, I only ever saw pointwise multiplication. Do you have an example of division you can share?

**Comment by kimihailv** [comment]
No, I don’t have such an example. I mentioned it because it’s possible that someone might do this in the future.

**Comment by sambhavnoobcoder** [comment]
hi @BenjaminBossan , 
i found this issue and the proposed solution quite interesting so i implemented it in #2825 . i would request you to check it out . i will look into any changes you have to suggest and will implement those as soon as possible as well .
